### DISCUSSION TITLE
:shield:  HTTPFS > Add headers (Access-Token:XXXX) :key:

### DISCUSSION THREAD

**Discussion by adriens** [duckdb/duckdb#6782] [Ideas]
# :grey_question: About

[`HTTPFS` ](https://duckdb.org/docs/extensions/httpfs.html)is actually an awesome way to get data from APIs. But sometimes, we have to authenticate on them. So actually I have to  : 

Run a query, eg:

```shell
# Lister les formations
http GET https://any.api.com/data \
    Access-Token:XXXXXXXXXXXXXXXXXXXXXX
```

... then pipe it into `jq`

... finally load it into `duckdb`

I generaly pipe this : 

```shell
# Lister les utilisateurs
http GET https://any.api.com/data \
   Access-Token:XXXX |\
   jq -r '.utilisateurs[] | [.col1, .col2, .col3, .col4, .col5] | @csv'\
   > data.csv
```

# :bulb: Feature request

- Make it possible to add headers via  [configuration](https://duckdb.org/docs/extensions/httpfs.html#configuration) : 
 
![image](https://user-images.githubusercontent.com/5235127/226146328-6acea468-830e-4123-a5f1-7eaabf56ebc3.png)

**Comment by Tishj** [comment]
This has come up before, and I think Elliana (Mause) suggested that we turn this setting into a MAP, so you can define arbitrary key-value pairs that way.

**Reply by adriens** [reply]
This is a brilliant idea :star_struck:

**Comment by nicosuave** [comment]
If you're using Python, check [this out](https://github.com/duckdb/duckdb/discussions/4963#discussioncomment-6068626) for the time being

**Reply by cboettig** [reply]
Just joining in here to add that I'd also love the ability to set arbitrary headers as key-value pairs in httpfs.  I think my use case could be pretty general -- one of the fantastic things about `httpfs` interface is it's ability to query plain https files -- an ability that features prominently in `duckdb`'s own test suite of the httpfs interface by operating against small parquet files on GitHub.  Of course GitHub repos aren't a great place for big data files, but GitHub's package registry is (as nicely described in this [blog post](https://betterprogramming.pub/how-homebrew-serves-52m-packages-per-month-413b9f0cf685)).  Only to query against storage in GHCR, we need to set an authentication header to the anonymous value, `Authorization: Bearer QQ==`, e.g.


curl -fL --header "Authorization: Bearer QQ==" https://ghcr.io/v2/cboettig/content-store/blobs/sha256:871017a6573cf4cd1ae303644de6bd0b04a1029278bfc80c94c37f968687e6d4 -o test.parquet

It would be great if duckdb could be used to read arbitrary data files from the GHCR registry just like it can already from S3 buckets... all we need is one header value.

**Reply by adriens** [reply]
Yes, that was exactly the plan @cboettig  👍 💯

**Comment by Mause** [comment]
I've done some initial work on this feature here: https://github.com/duckdb/duckdb/pull/8425

**Comment by adriens** [comment]
:partying_face:  [Query an Authenticated API Endpoint](https://duckdbsnippets.com/snippets/184/query-an-authenticated-api-endpoint)

![image](https://github.com/user-attachments/assets/b6c7057c-b207-4175-83d7-369f4e67d69f)

**Comment by adriens** [comment]
- https://github.com/duckdb/duckdb/discussions/14165

**Comment by carlopi** [comment]
I think this is solved via EXTRA_HTTP_HEADER Map!

**Reply by adriens** [reply]
Indeed yes :+1: :partying_face:
